In [ ]:
!pip install pysam
import pysam

Reading BAM file

In [74]:
samfile = pysam.AlignmentFile("merged-tumor.bam","rb")

Fetching first read and inspecting it

In [75]:
read=samfile.head(1)
flag=0
for x in read:
  print(x.to_string())
  flag=x.flag

C0HVYACXX120402:7:1207:5722:57044	1187	21	9483249	27	76M	=	9483382	209	TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG	==<>@?@@>@D?>@C?>8JA5?>CC>?@???BB;H-:4A?D=DB8BDE?GBJCD@B8??E<A>CDJB@B>AA@@@C	XA:Z:GL000217.1,-110754,76M,1;	BD:Z:NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN	MD:Z:76	RG:Z:1	BI:Z:NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN	NM:i:0	MQ:i:27	AS:i:76	XS:i:71


Inspecting flag field:

In [76]:
print(flag)

1187


Result from https://broadinstitute.github.io/picard/explain-flags.html for flag 1187:
Summary:
    read paired (0x1)
    read mapped in proper pair (0x2)
    mate reverse strand (0x20)
    second in pair (0x80)
    read is PCR or optical duplicate (0x400)

In [77]:
unmapped_reads=0
total_reads=0
map_quality_zero=0
sum_all=0
for read in samfile:
  flag=read.flag
  total_reads+=1
  if(flag&0x4):
    unmapped_reads+=1
  if(read.mapping_quality==0):
    map_quality_zero+=1
  sum_all+=read.mapping_quality
print("Number of unmaped reads: ",unmapped_reads)
print("Total number of reads: ",total_reads)
print("Number of reads with mapping quality 0: ",map_quality_zero)
avg_quality=sum_all/(total_reads*1.0)
avg_non_zero=sum_all/((total_reads-map_quality_zero)*1.0)
print("Average mapping quality of all reads: ",avg_quality)
print("Average mapping quality (reads with zero mapping quality filtered out): ",avg_non_zero)

Number of unmaped reads:  17765
Total number of reads:  2921629
Number of reads with mapping quality 0:  126628
Average mapping quality of all reads:  55.91379158681681
Average mapping quality (reads with zero mapping quality filtered out):  58.4469755109211
